### PR TITLE
Preserve state of placeholders by hiding timed-out children instead of deleting them

### PR DESCRIPTION
**Pull request by acdlite**
When a placeholder times out, instead of unmounting the current tree, hide it and render the fallback view as a sibling. When the data resolves and the tree resumes rendering, the children are made visible again. Because they remain mounted throughout, the state is preserved.

Adds additional host config methods. For mutation mode:

- hideInstance
- hideTextInstance
- unhideInstance
- unhideTextInstance

For persistent mode:

- cloneHiddenInstance
- cloneUnhiddenInstance
- createHiddenTextInstance

### TODO

- [x] Need more comments.
- [ ] Implement new host config methods for each renderer:
  - [x] Noop
  - [x] DOM
  - [x] ART
  - [ ] Native
  - [ ] Fabric
  - [x] Test
    - Idk how this should be exposed publicly yet so all I did was add an `isHidden` property to the instance types.

### CODE REVIEW

**Comment by react-linter** [comment]
ReactDOM: size: :small_red_triangle:+0.9%, gzip: :small_red_triangle:+0.9%


  <details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: baff5cc2f69d30589a5dc65b089e47765437294b...9226bad4a8dfb32d3c7f8149f5d60201556fcd5c</p>


  
## react-dom
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-dom.development.js | +2.3% | +2.6% | 627.88 KB | 642.25 KB | 146.85 KB | 150.61 KB | UMD_DEV
**react-dom.production.min.js** | **:small_red_triangle:+0.9%** | **:small_red_triangle:+0.9%** | **95.52 KB** | **96.34 KB** | **30.85 KB** | **31.13 KB** | **UMD_PROD**
react-dom.development.js | +2.3% | +2.6% | 624 KB | 638.38 KB | 145.67 KB | 149.44 KB | NODE_DEV
**react-dom.production.min.js** | **:small_red_triangle:+0.9%** | **:small_red_triangle:+0.9%** | **95.51 KB** | **96.34 KB** | **30.39 KB** | **30.67 KB** | **NODE_PROD**
ReactDOM-dev.js | +2.4% | +2.7% | 630.02 KB | 645.31 KB | 143.94 KB | 147.84 KB | FB_WWW_DEV
**ReactDOM-prod.js** | **:small_red_triangle:+2.7%** | **:small_red_triangle:+2.4%** | **272.46 KB** | **279.75 KB** | **51.33 KB** | **52.56 KB** | **FB_WWW_PROD**
react-dom.profiling.min.js | +0.9% | +0.9% | 96.48 KB | 97.31 KB | 30.73 KB | 31.01 KB | NODE_PROFILING

## react-art
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-art.development.js | +3.6% | +4.2% | 411.86 KB | 426.86 KB | 92.41 KB | 96.29 KB | UMD_DEV
**react-art.production.min.js** | **:small_red_triangle:+1.1%** | **:small_red_triangle:+1.3%** | **82.66 KB** | **83.54 KB** | **25.45 KB** | **25.78 KB** | **UMD_PROD**
react-art.development.js | +4.4% | +5.2% | 343.84 KB | 358.83 KB | 75.27 KB | 79.17 KB | NODE_DEV
**react-art.production.min.js** | **:small_red_triangle:+1.9%** | **:small_red_triangle:+2.1%** | **47.64 KB** | **48.53 KB** | **14.83 KB** | **15.15 KB** | **NODE_PROD**
ReactART-dev.js | +4.8% | +5.7% | 332.67 KB | 348.6 KB | 70.01 KB | 74.01 KB | FB_WWW_DEV
**ReactART-prod.js** | **:small_red_triangle:+5.3%** | **:small_red_triangle:+4.9%** | **143.39 KB** | **150.97 KB** | **24.69 KB** | **25.9 KB** | **FB_WWW_PROD**

## react-test-renderer
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-test-renderer.development.js | +4.3% | +5.1% | 348.35 KB | 363.35 KB | 75.83 KB | 79.72 KB | UMD_DEV
**react-test-renderer.production.min.js** | **:small_red_triangle:+1.8%** | **:small_red_triangle:+2.1%** | **48.53 KB** | **49.41 KB** | **14.89 KB** | **15.21 KB** | **UMD_PROD**
react-test-renderer.development.js | +4.4% | +5.2% | 344.47 KB | 359.47 KB | 74.85 KB | 78.75 KB | NODE_DEV
**react-test-renderer.production.min.js** | **:small_red_triangle:+1.9%** | **:small_red_triangle:+2.3%** | **48.24 KB** | **49.14 KB** | **14.71 KB** | **15.04 KB** | **NODE_PROD**
react-test-renderer-shallow.development.js | +0.1% | 0.0% | 21.64 KB | 21.65 KB | 5.9 KB | 5.91 KB | UMD_DEV
**react-test-renderer-shallow.production.min.js** | **:small_red_triangle:+0.1%** | **0.0%** | **7.16 KB** | **7.16 KB** | **2.34 KB** | **2.34 KB** | **UMD_PROD**
react-test-renderer-shallow.development.js | +0.1% | 0.0% | 16.83 KB | 16.84 KB | 4.61 KB | 4.61 KB | NODE_DEV
**react-test-renderer-shallow.production.min.js** | **:small_red_triangle:+0.1%** | **0.0%** | **7.77 KB** | **7.78 KB** | **2.58 KB** | **2.58 KB** | **NODE_PROD**
ReactTestRenderer-dev.js | +4.6% | +5.5% | 347.69 KB | 363.63 KB | 73.58 KB | 77.61 KB | FB_WWW_DEV
ReactShallowRenderer-dev.js | +0.1% | +0.1% | 16.46 KB | 16.48 KB | 4.29 KB | 4.3 KB | FB_WWW_DEV

## react-noop-renderer
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-noop-renderer.development.js | +8.1% | +3.7% | 18.07 KB | 19.53 KB | 4.72 KB | 4.9 KB | NODE_DEV
**react-noop-renderer.production.min.js** | **:small_red_triangle:+7.5%** | **:small_red_triangle:+5.4%** | **6.71 KB** | **7.21 KB** | **2.58 KB** | **2.72 KB** | **NODE_PROD**
react-noop-renderer-persistent.development.js | +8.1% | +3.6% | 18.19 KB | 19.66 KB | 4.74 KB | 4.91 KB | NODE_DEV
**react-noop-renderer-persistent.production.min.js** | **:small_red_triangle:+7.5%** | **:small_red_triangle:+5.5%** | **6.73 KB** | **7.23 KB** | **2.58 KB** | **2.73 KB** | **NODE_PROD**

## react-reconciler
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-reconciler.development.js | +4.4% | +5.2% | 333.16 KB | 347.66 KB | 71.39 KB | 75.13 KB | NODE_DEV
**react-reconciler.production.min.js** | **:small_red_triangle:+1.9%** | **:small_red_triangle:+2.0%** | **47.16 KB** | **48.05 KB** | **14.13 KB** | **14.41 KB** | **NODE_PROD**
react-reconciler-persistent.development.js | +4.4% | +5.3% | 331.78 KB | 346.27 KB | 70.83 KB | 74.58 KB | NODE_DEV
**react-reconciler-persistent.production.min.js** | **:small_red_triangle:+1.9%** | **:small_red_triangle:+2.0%** | **47.17 KB** | **48.06 KB** | **14.14 KB** | **14.42 KB** | **NODE_PROD**
react-reconciler-reflection.development.js | +0.1% | 0.0% | 14.09 KB | 14.1 KB | 4.43 KB | 4.43 KB | NODE_DEV

## react-native-renderer
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
ReactNativeRenderer-dev.js | +3.3% | +3.8% | 466.86 KB | 482.11 KB | 102.79 KB | 106.7 KB | RN_FB_DEV
**ReactNativeRenderer-prod.js** | **:small_red_triangle:+3.6%** | **:small_red_triangle:+3.8%** | **207.33 KB** | **214.78 KB** | **36.52 KB** | **37.9 KB** | **RN_FB_PROD**
ReactNativeRenderer-dev.js | +3.3% | +3.8% | 466.57 KB | 481.82 KB | 102.73 KB | 106.64 KB | RN_OSS_DEV
**ReactNativeRenderer-prod.js** | **:small_red_triangle:+1.5%** | **:small_red_triangle:+2.0%** | **199.61 KB** | **202.65 KB** | **34.98 KB** | **35.7 KB** | **RN_OSS_PROD**
ReactFabric-dev.js | +3.3% | +3.9% | 457.12 KB | 472.37 KB | 100.38 KB | 104.29 KB | RN_FB_DEV
**ReactFabric-prod.js** | **:small_red_triangle:+1.6%** | **:small_red_triangle:+1.8%** | **191.87 KB** | **194.97 KB** | **33.57 KB** | **34.19 KB** | **RN_FB_PROD**
ReactFabric-dev.js | +3.3% | +3.9% | 457.15 KB | 472.4 KB | 100.4 KB | 104.31 KB | RN_OSS_DEV
**ReactFabric-prod.js** | **:small_red_triangle:+1.6%** | **:small_red_triangle:+1.8%** | **191.9 KB** | **195 KB** | **33.58 KB** | **34.2 KB** | **RN_OSS_PROD**
ReactNativeRenderer-profiling.js | +1.5% | +2.0% | 202.29 KB | 205.32 KB | 35.58 KB | 36.31 KB | RN_OSS_PROFILING
ReactFabric-profiling.js | +1.6% | +1.9% | 194.26 KB | 197.35 KB | 34.1 KB | 34.75 KB | RN_OSS_PROFILING

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: packages/react-dom/src/client/ReactDOMHostConfig.js [hunk 1]
```diff
@@ -394,6 +399,41 @@ export function removeChildFromContainer(
   }
 }
 
+export function hideInstance(instance: Instance): void {
+  // TODO: Does this work for all element types? What about MathML? Should we
+  // pass host context to this method?
+  instance = ((instance: any): HTMLElement);
+  if (instance.style !== undefined && instance.style !== null) {
+    instance.style.display = 'none';
```
Comment (gaearon) [inline packages/react-dom/src/client/ReactDOMHostConfig.js]: What happens if it already has a style attribute that disagrees? What if you hide it first, but then it updates itself with `block` as the value? Or does sequencing guarantee `hideInstance` always runs last?

Comment (acdlite) [reply packages/react-dom/src/client/ReactDOMHostConfig.js]: During the commit that hides them, `hideInstance` always comes last because children effects fire before parent effects. While they're timed out, the children will never update. It's a more extreme version of `hidden`.

File: packages/react-reconciler/src/ReactFiberBeginWork.js [hunk 2]
```diff
@@ -742,25 +739,106 @@ function updatePlaceholderComponent(
...
+    // fallbackChildren]. We always render both slots, to avoid reconciliation
+    // mismatches, but a non-timed-out placeholder's fallback children are
+    // always null.
+    //
+    // When the placeholder times out, the normal children are not unmounted.
+    // We keep them in the tree to preserve their existing state. Then the
+    // renderer hides them visually e.g. using display: none.
+    //
+    // The fallback children are wrapped in an extra fiber with a special,
+    // private type. We use the type to distinguish between the two sets.
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberBeginWork.js]: @sebmarkbage This is the most important part to review. I'm not that happy with this approach; I think the ideally we would store each set of children on a separate field, the way call/return used to store its first child set on the state node. I resisted that at first because I thought it would be cumbersome to have to account for both sets every time we traverse the fiber tree (including context). But it turns out that I have to account for it in most of the places regardless.

Any alternative ideas?

**Comment by acdlite** [comment]
This is also a whole lot of code for something that's supposed to be an edge case (an already mounted component causes an update to time out). Previous discussions indicated this was the least bad option, but maybe if we try harder we can think of a better trade off.

File: packages/react-reconciler/src/ReactFiberCompleteWork.js [hunk 3]
```diff
@@ -314,6 +316,67 @@ function completeWork(
 ): Fiber | null {
   const newProps = workInProgress.pendingProps;
 
+  // Reset the child expiration time by bubbling it from the children. Skip
+  // if the children are hidden and we're not rendering at hidden priority.
```
Comment (sebmarkbage) [inline packages/react-reconciler/src/ReactFiberCompleteWork.js]: There are a number of little fixes and refactoring in this commit that isn't directly related to the hiding of children. Such as moving this section and the extra arguments and move of retrySuspendedRoot.

It would be nice to have a separate PR for those since they seem like useful clean up regardless.

File: packages/react-reconciler/src/ReactFiberCompleteWork.js [hunk 4]
```diff
@@ -84,17 +87,89 @@ function markRef(workInProgress: Fiber) {
   workInProgress.effectTag |= Ref;
 }
 
-function appendAllChildren(parent: Instance, workInProgress: Fiber) {
+function appendAllChildren(
+  parent: Instance,
+  workInProgress: Fiber,
+  needsVisibilityToggle: boolean,
+  isHidden: boolean,
+) {
```
Comment (sebmarkbage) [inline packages/react-reconciler/src/ReactFiberCompleteWork.js]: It's probably worth just forking this function between the two modes.

File: packages/react-art/src/ReactARTHostConfig.js [hunk 5]
```diff
@@ -407,21 +407,22 @@ export function commitUpdate(
...
+  // Noop
 }
 
-export function unhideInstance(instance: Instance, props: Props): void {
-  throw new Error('Not yet implemented.');
+export function unhideInstance(instance, props) {
+  if (props.visible == null || props.visible) {
+    instance.show();
+  } else {
+    instance.hide();
```
Comment (sebmarkbage) [inline packages/react-art/src/ReactARTHostConfig.js]: This is probably unnecessary since it's always hidden already then.

File: packages/react-reconciler/src/ReactFiberBeginWork.js (hunk 2 above)
Comment (sebmarkbage) [reply packages/react-reconciler/src/ReactFiberBeginWork.js]: What if the stateNode was null when it is not timed out. We store normalChildren on firstChild. Then when we timeout, we swap the normalChildren and store the normalChildren on the stateNode while it is timed out and the fallbackChildren on firstChild.

Whenever we retry to see if we're unblocked we try to restore the normalChildren to firstChild first and then render down. If we don't succeed, we restore them to the placeholder again.

File: packages/react-dom/src/client/ReactDOMHostConfig.js (hunk 1 above)
Comment (NE-SmallTown) [reply packages/react-dom/src/client/ReactDOMHostConfig.js]: This makes me remember [this](https://github.com/NE-SmallTown/Thinking-in-react/tree/master/packages/react-suspense-polyfill) and [this](https://twitter.com/Heaven_xz/status/972139052748062720) :)

**Comment by acdlite** [comment]
Closing in favor of #13823
